Working proposal for adding RSQLite (SQLite/json1)

rfhb opened the pull request:
This is a PR to supersede #24, addressing review comments and extending functionality. 

Included is a working implementation of nodbi for use with RSQLite, in analogy to existing methods (in particular MongoDB). Also, a nested ragged json dataset (`contacts`) is included for testing. 

RSQLite ships since 2016 with an sqlite version that includes the json1 extension, thus enabling a JSON store for small projects or for exploring. 

The query method has many lines, for translating json strings for `query` and `fields` into SQL.  

Thanks

sckott commented:
thanks @rfhb sorry for the delay, vacation and other work stuff. 

Will have a look at this hopefully today

rfhb commented:
Thanks. Am still working on the query method to make this work exactly like the mongo method. Perhaps I can ping you here once the additional works are ready.

sckott commented on R/create.R:
```diff
@@ -33,6 +33,12 @@
+#' # SQLite
+#' src <- src_sqlite()
+#' docdb_create(src, key = "mtcars", value = mtcars)
+#' docdb_create(src, key = "mtcars", value = contacts)
```
Currently, the package is setup to work with data.frame's - so e.g., the example with the contacts dataset fails

I think we'd prefer to keep the pkg as simple as possible and not try to support other data types

rfhb replied:
Many thanks @sckott for review and suggestions! 
Ok - I have changed the proposal to work exclusively with data.frames. 
When a data.frame contains JSON strings, these are specially processed.

sckott commented on R/get.R:
```diff
@@ -91,10 +98,47 @@ docdb_get.src_mongo <- function(src, key, limit = NULL, ...) {
+  dump <- tempfile()
+  
+  # get data, write to file in ndjson format
+  cat(stats::na.omit(unlist(
```
I wonder if this is necessary to write to a file first? I haven't looked at this in detail so I may be wrong here. I guess `jsonlite::stream_in` only can accept a file/url/gzcon, so maybe file is the only way

rfhb replied:
In the code comments just before `jsonlite::stream_in()`, I cite the jsonlite developer's reason for this function: parsing huge JSON strings is difficult and inefficient (see also [https://rdrr.io/cran/jsonlite/man/stream_in.html#heading-3](https://rdrr.io/cran/jsonlite/man/stream_in.html#heading-3). 

In addition, `nodbi::docdb_get.src_mongo()` already used `jsonlite::stream_in()` so that I considered this a relevant option also for `nodbi::docdb_get.src_sqlite()`. The file in a temporary space of the local file system greatly simplifies implementation. 

Also it comes with a significant speed improvement, as can be seen now for `nodbi::docdb_create.src_sqlite()` and this addresses the issue raised above.

sckott replied:
ok, thanks

sckott commented on NAMESPACE:
```diff
@@ -5,30 +5,40 @@ S3method(docdb_create,src_elastic)
 S3method(docdb_exists,src_couchdb)
 S3method(docdb_exists,src_elastic)
 S3method(docdb_exists,src_etcd)
+S3method(docdb_exists,src_mongo)
```
remove those mongo changes in the NAMESPACE

rfhb replied:
Corrected, thank you.

sckott commented:
any thoughts @rfhb ?

rfhb commented:
Many thanks @sckott for getting back - I had been working on this over the last weeks. Happy to hear back from your thoughts on the latest commits!

sckott commented:
thanks @rfhb - can you address my question about performance?

rfhb commented:
thanks @sckott - performance is well improved, mentioned this in my comment https://github.com/ropensci/nodbi/pull/25#discussion_r293123757 and here are figures on a laptop: 

```
library(nodbi)

## persistent database
unlink("test.sqlite")
src <- src_sqlite("test.sqlite")
system.time(docdb_create(src, "diamonds", diamonds))
#> 20 sec elapsed time

## memory-based database
src <- src_sqlite()
system.time(docdb_create(src, "diamonds", diamonds))
#> 0.8 sec elapsed time
```

The difference in execution time is solely due to `RSQLite::dbWriteTable()` or its generic function `DBI::dbAppendTable()`, which takes longer for writing to disk compared to memory.

sckott commented:
thanks @rfhb - leaving on vacation for a week - will take another look then - sorry about the delay

rfhb commented:
Hi, would you want to have another look at the PR, please? I am now starting to use it for another project and believe to have addressed comments so far. Many thanks.

sckott commented:
having a look now

sckott commented:
You're right, the performance is very good in memory, and much improved when writing to disk.

sckott commented:
looks good, thanks for your contribution, assigned on next milestone https://github.com/ropensci/nodbi/milestone/3
